File: 1일1알고/dfs_bfs/1227/2468.py
import sys
sys.stdin = open('2468.txt')

from collections import deque


# def findMax(N):
#     global mmax
#     for i in range(N):
#         for j in range(N):
#             if map[i][j]>=mmax:
#                 mmax=map[i][j]
#     return mmax
#
#
#
# def bfs(i,j,height,visited):
#     global cnt
#     queue = deque()
#     queue.append((i,j))
#
#     visited[i][j] = 1
#
#     while queue:
#         i,j = queue.popleft()
#         for di,dj in [[0,1],[1,0],[0,-1],[-1,0]]:
#             ni,nj = i+di, j+dj
#             if 0<=ni<N and 0<=nj<N and map[ni][nj] > height and visited[ni][nj] == 0:
#                 visited[ni][nj] = 1
#                 queue.append((ni,nj))
#
#
#
#
#
#
# N = int(input())
# map = [list(map(int,input().split())) for _ in range(N)]
# mmax = 0
# maxHeight = findMax(N)
# # print(maxHeight)
#
# #maxHieght까지 안전영역을 찾으면서 반복문으로 하나씩 높이면서 찾으면서
# #최대갯수를 갱신
#
# #반복문 안에서
# #높이마다 안전영역 세팅
# max_ans = 0
#
# for height in range(1,maxHeight+1):
#     visited = [[0]* N for i in range(N)]
#
#     cnt = 0
#     for i in range(N):
#         for j in range(N):
#             if map[i][j] > height and visited[i][j] == 0: #더 높은 위치고 방문안했으면 안전영역이면
#                 bfs(i,j,height, visited)
#                 cnt += 1
#
#
#     if max_ans<=cnt:
#         max_ans=cnt
#
# print(max_ans)




def findMax(N):
    x = 0 #전역변수를 쓸 이유가 전혀없지요
    for i in range(N):
        for j in range(N):
            if arr[i][j]>x:
                x=arr[i][j]
    return x



def bfs(i,j):
    queue = deque()
    queue.append((i,j))

    visited[i][j] = 1

    while queue:
        i,j = queue.popleft()
        for di,dj in [[0,1],[1,0],[0,-1],[-1,0]]:
            ni,nj = i+di, j+dj
            if 0<=ni<N and 0<=nj<N and arr[ni][nj] > height and visited[ni][nj] == 0: #안전영역이고 방문한적 없으면 #하나의 안전영역 안에서 중복검사 방지
                visited[ni][nj] = 1
                queue.append((ni,nj))






N = int(input())
arr = [list(map(int,input().split())) for _ in range(N)]
maxHeight = findMax(N)
# print(maxHeight)

#maxHieght까지 안전영역을 찾으면서 반복문으로 하나씩 높이면서 찾으면서
#최대갯수를 갱신

#반복문 안에서
#높이마다 안전영역 세팅
max_ans = 0

for height in range(maxHeight+1):
    visited = [[0]* N for i in range(N)]

    cnt = 0
    for i in range(N):
        for j in range(N):
            if arr[i][j] > height and visited[i][j] == 0: #더 높은 위치고 방문안했으면 안전영역이면 #안전영역 갯수 늘어날때마다 이미 검사한 안전영역 못하게
                bfs(i,j)
                cnt += 1


    if max_ans<=cnt:
        max_ans=cnt

print(max_ans)

File: 1일1알고/dfs_bfs/1227/2468.js
let input = require("fs")
  .readFileSync("2468.txt")
  .toString()
  .trim()
  .split("\n");

const N = +input[0];
const arr = [];
for (let i = 1; i < N + 1; i++) {
  arr.push(input[i].trim().split(" ").map(Number));
}



const findMax = (N) => {
    let mmax = 0
    for (let i=0; i<N; i++) {
        for (let j=0; j<N; j++) {
            if (mmax < arr[i][j]) {
                mmax = arr[i][j]
            }
        }
    }
    return mmax
}

// const bfs = (graph, start, visited) => {
//     const q = [];
//     q.push(start);
//     visited[start] = true;
  
//     while (q.length !== 0) {
//       const v = q.shift();
//       console.log(v);
      
//       for(const cur of graph[v]){
//         if(!visited[cur]){
//           q.push(cur);
//           visited[cur] = true;
//         }
//       }
//     }
//   }
const bfs = (i,j,visited, height) => {
  const q = []
  q.push([i,j])
  visited[i][j] = true;
  const dir = [[0,1],[1,0],[0,-1],[-1,0]]

  while ( q.length !== 0 ) {
    const [i,j] = q.shift()
    for (let k=0; k<4; k++) {
      let ni = i+dir[k][0]
      let nj = j+dir[k][1]
      if (ni>=0 && ni<N && nj>=0 && nj<N && arr[ni][nj] > height && visited[ni][nj] === false) {
        visited[ni][nj] = true
        q.push([ni,nj])
      } 
    }
  }
}

const maxHeight = findMax(N)

let max_ans = 0

for (let height=0; height<maxHeight+1; height++) {
  let visited = Array.from(Array(N), ()=> Array(N).fill(false))
  let cnt = 0
  for (let i=0;i<N;i++) {
    for (let j=0;j<N;j++) {
      if ( arr[i][j] > height && visited[i][j] === false ) {
        bfs(i,j,visited,height)
        cnt += 1
      }
    }
  }
  if (max_ans<=cnt) {
    max_ans=cnt
  }
}


console.log(max_ans)
